### ISSUE TITLE
v0.9.8 - 'Unlink' never removes previous link relation changes.

### ISSUE THREAD

**Issue by lp-belliot** [maritz/nohm#122]
Love the package! It's been very helpful of late, but I've noticed an issue :)

I noticed that if you persist a parent model in memory, the relationChanges array just continues to grow as you link and unlink objects to the parent object.

It looks like there is a check failing in the unlink function for a couple of reasons.

In source, the function looks like this:
```js
exports.unlink = function unlink(obj, options, callback) {

  if (typeof(options) === 'string') {
    options = {name: options};
  }
  var opts = h.$extend({
    name: 'default'
  }, options);

  callback = h.getCallback(arguments);
  
  var changes = this.relationChanges;
  for (var i in changes) {
    if (changes.hasOwnProperty(i) &&
        changes[i].name === opts.name &&
        h.checkEqual(changes[i].object, obj)) {
      delete this.relationChanges[i];
    }
  }
  
  this.relationChanges.push({
    action: 'unlink',
    options: opts,
    callback: callback,
    object: obj
  });
};
```

First, `changes[i].name === opts.name` will always evaluate false, as the relation name is actually at `changes[i].options.name`.

Second, `delete this.relationchanges` will leave an 'undefined' object in the array at the index, causing issues further downstream. You're better off using `this.relationChanges.splice(i, 1);`

Third, in the `h.checkEqual()` function, there is a conditional as below:
```js
else if (obj1.hasOwnProperty('modelName') && obj2.hasOwnProperty('modelName') &&
           obj1.modelName === obj2.modelName) {
```
This will exclude the entire second half of the function, as the modelName attribute is located at `obj1.__proto__.modelName`

I'll prepare a PR with the appropriate changes to fix the issue. I'm not fussed if you would rather approach the issue in a different manner, but we're keen to get this merged to avoid issues with patching at build.

**Comment by lp-belliot** [comment]
I've put the changes here:

https://github.com/lp-belliot/nohm/tree/feature/fix_bug_in_unlink_function

Not sure what to do given master is well past the v0.9.8 commit.

**Comment by sam2x** [comment]
Let me check this, I think this is same bug I got before. @maritz if that's the case, and the PR works well, is there any change to get it merged to the v1 ? Most users still uses it (I haven't also migrated yet to v2).

**Comment by maritz** [comment]
Sure, I just won't spend much time on it in v1.  ;-)  
But I can certainly read, accept/decline PRs and publish new versions.

The checkEqual is still the same in v2, the other 2 problems are already fixed in v2.

@lp-belliot thanks for the report & PR! Can you add some unit tests that fail without your changes? If possible one for each of the 3 problems you found.

**Comment by lp-belliot** [comment]
Thanks!

I've added some unit tests to the branch linked above.

Let me know which branch to point the PR at if you're happy with the changes!

**Comment by maritz** [comment]
@lp-belliot Branch for that doesn't exist yet, I'll make one later.

@sam2x I guess this means we'll have a v1 maintenance branch. Would you be interested in maintaining that? This would probably also mean occasionally updating dependencies and making sure things still work afterwards.  

I'm not keen on doing that work, that's why I'm trying to emphasize v2. However I have also been held up on completing v2, so I fully understand that no one can easily use it yet.

**Comment by maritz** [comment]
@lp-belliot I have created the https://github.com/maritz/nohm/tree/v1_maintenance branch that you can point the PR at.

I will cherry-pick the helper changes and tests for v2 as well, if you don't mind.

**Comment by maritz** [comment]
Fixed in 2.0.0-alpha.12 and 1.1.0-breaking.

To use it while keeping your code on the old v1 branch, use the [v1_breaking_maintenance branch](https://github.com/maritz/nohm/blob/v1_breaking_maintenance/) via npm tag:

```bash
npm install nohm@v1-breaking
```

or by using the specific version:

```bash
npm install nohm@1.1.0-breaking
```
